### ISSUE TITLE
question: safe/completed

### ISSUE THREAD

**Issue by immesys** [ceph/go-ceph#28]
I can't find in the librados documentation (or the ceph-go documentation) when in the process the non-aio writes return. Is the data in memory on a majority of the pg, in memory on all of the pg, or on disk on all of the pg?

Thanks

**Comment by noahdesu** [comment]
does the following answer your question?

```
 * @param cb_complete the function to be called when the operation is
 * in memory on all relpicas
 * @param cb_safe the function to be called when the operation is on
```

https://github.com/ceph/ceph/blob/master/src/include/rados/librados.h#L1745

**Comment by immesys** [comment]
I'm not sure. It does if you are implying that the synchronous writes use the aio_completions under the covers. Perhaps let me rephrase my question. I am used to using the aio functions which have clearly documented behavior regarding the consistency. The go bindings use the normal (not aio) write functions, which are not as clearly documented. When I call Write() on the go IOContext, and it returns, at that stage is that equivalent to the aio safe, the aio complete or something else?

**Comment by noahdesu** [comment]
I see. The normal synchronous writes will be the same as waiting for the "safe" state with AIO.

**Comment by immesys** [comment]
perfect, thanks
